File: Lab1_SecretKeyCryptography/Solutions/task_01/ngram_score.py
'''
    ngram_score module to provide a mechanism to score
    texts based on a n-gramm lookup table. The lookup
    table has to be created out-of-band.

    Date 11.10.2021
'''

from math import log10



'''
    ngram_score class to calculate the n-gramm score
    of a text based on a lookup table of the most
    common n-gramms in a specific language.
'''
class ngram_score(object):

    def __init__(self, file_name, sep=' '):
        '''
            Construct a new n-gram lookup table from the
            provided file. The assumed file structure is
            <ngram> <number-of-occurance>, separated by
            a whitespace.
        '''

        # reak in raw file
        self.ngrams = {}
        for line in open(file_name, 'r'):
            ngram, count = line.split(sep)
            self.ngrams[ngram] = int(count)
        
        # fix some internal parameters
        self.order = len(ngram)
        self.total_ngrams = sum(self.ngrams.values())

        # calculate propabilities
        for ix in self.ngrams.keys():
            p = log10(float(self.ngrams[ix]) / self.total_ngrams)
            self.ngrams[ix] = p

        # define default propability for n-gramms
        # not occuring in the given lookup table
        self.default_value = log10(0.01/self.total_ngrams)


    def order(self):
        '''
            Return the order of the n-gramm.
        '''
        return self.order


    def score(self, input_text, normalize=False):
        '''
            Calculate the Score of the input text based on
            the lookup table. The option 'normalize' is uesed
            to normalize the score based on the text input
            length. While this is required to compare different
            length texts, it has negative effects on scoring
            texts of same length! Only enable it if required!
        '''
        score = 0
        text = input_text.upper()
        for idx in range(len(text)-self.order+1):
            current_ngram = text[idx:idx+self.order]
            if current_ngram in self.ngrams:
                score += self.ngrams[current_ngram]
            else:
                score += self.default_value

        # normalize the score to the length of the
        # given input text. Otherwise longer texts
        # will automatically have lower score, even
        # if they were better natural language
        if normalize:
            score = score / (len(text)-self.order+1)

        return score
    

File: Lab1_SecretKeyCryptography/Solutions/task_01/ngram_score.cpp
#include "ngram_score.h"

#include<iostream>
#include<fstream>
#include<cmath>
#include<algorithm>
#include<numeric>


/**
 * Constructs a new ngram_score object from the
 * ngram_file providing the list of n-grams and
 * there related occurencies.
 */
ngram_score::ngram_score(const std::string& ngram_file)
    : gram_order(-1), default_propability(0.0f)
{
    // read in the n-gramms from file to STL map and count
    // the total number of read in data for normalization
    std::ifstream file(ngram_file.c_str());
    std::string key;
    float value = 0.0f;
    size_t total_num_ngrams = 0;
    while(file >> key >> value)
    {
        n_gramms[key] = value;
        total_num_ngrams += value;
        
    }

    // fix some internal parameters
    // I.e., order of the read n-gramms, total number read entities
    gram_order = n_gramms.empty() ? 0 : n_gramms.begin()->first.length();

    // normalize occurencies to log scale propability
    for(std::map<const std::string, float>::iterator iter = n_gramms.begin(); 
        iter != n_gramms.end(); ++iter)
    {
        iter->second = std::log10(iter->second / total_num_ngrams);
    }

    // Set default value for all n-grams may later processed by get_score()
    // which are not in our look-up table
    default_propability = std::log10(0.01f / static_cast<float>(total_num_ngrams));
}


 /**
   * Calculate the Score of the input text based on
   * the lookup table.
   */
const float ngram_score::get_score(const std::string& input_text, const bool normalize) const
{
    // transform input string to upper case
    std::string text(input_text);
    std::transform(text.begin(), text.end(),text.begin(), ::toupper);
    std::cout << text << std::endl;
    float score = 0.0f;
    const size_t num_grams = text.length() - gram_order + 1;
    for(size_t ix = 0; ix != num_grams; ++ix)
    {
        const std::string current_gram = text.substr(ix, ix + gram_order);
        if( n_gramms.find(current_gram) != n_gramms.end() )
        {
            score += n_gramms.at(current_gram);
        }
        else 
        {
            score += default_propability;
        }
    }

    // normalize the score to the length of the
    // given input text. Otherwise longer texts
    // will automatically have lower score, even
    // if they were better natural language
    if(normalize)
    {
        score /= (text.length() - gram_order + 1);
    }

    return score;
}
